### PR TITLE
FFVA 48KHz USB buffer overflow fix

### PR DESCRIPTION
**Pull request by shuchitak**
https://xmosjira.atlassian.net/browse/AP-153

I've ported the increased precision USB rate calculation changes from the ASRC demo example to the FFVA example. 

Without the fix, with what's on develop, the samples_to_host_stream_buf level looks like this, where the y axis is the number of unread samples in the buffer -
ffva_adaptive_usb_orig
The buffer gets full and is reset in the driver (the sharp drop from 1000 something samples to 0) about every ~ 16 seconds.

After porting the USB rate calculation logic from ASRC app to FFVA, the buffer overflow issue is not seen. I ran for 10 mins on my Mac and the buffer level remained constant.
stable_buffer_ffva

We could also add a closed loop control based on the buffer level but since I don't notice any drift, I'm not going to add it for now. Keeping this in mind as an option though.

I've also added support to run the adaptive algorithm for the EP_IN direction in case EP_OUT interface is not active.

### CODE REVIEW

File: examples/ffva/src/main.c [hunk 1]
```diff
@@ -312,7 +312,7 @@ void vApplicationMallocFailedHook(void)
 static void mem_analysis(void)
 {
 	for (;;) {
-		rtos_printf("Tile[%d]:\n\tMinimum heap free: %d\n\tCurrent heap free: %d\n", THIS_XCORE_TILE, xPortGetMinimumEverFreeHeapSize(), xPortGetFreeHeapSize());
+		//rtos_printf("Tile[%d]:\n\tMinimum heap free: %d\n\tCurrent heap free: %d\n", THIS_XCORE_TILE, xPortGetMinimumEverFreeHeapSize(), xPortGetFreeHeapSize());
```
Comment (ACascarino) [inline examples/ffva/src/main.c:315]: Would rather delete commented lines, or declare a debug unit for debug_printf that disables these printouts... can we do that for rtos_printf? unsure

File: examples/ffva/src/usb/adaptive_rate_adjust.c [hunk 2]
```diff
@@ -23,21 +23,27 @@
...
 
 typedef struct usb_audio_rate_packet_desc {
     uint32_t cur_time;
     uint32_t ep_num;
     uint32_t ep_dir;
     size_t xfer_len;
+    bool calc_rate;
 } usb_audio_rate_packet_desc_t;
 
+extern volatile bool mic_interface_open;
```
Comment (ACascarino) [inline examples/ffva/src/usb/adaptive_rate_adjust.c:42]: externed from where?

File: examples/ffva/src/usb/adaptive_rate_callback.c [hunk 3]
```diff
@@ -83,22 +50,68 @@ uint32_t sum_array(uint32_t * array_to_sum, uint32_t array_length)
     return acc;
 }
 
-void reset_state()
+
+float_s32_t float_div(float_s32_t dividend, float_s32_t divisor)
```
Comment (ACascarino) [inline examples/ffva/src/usb/adaptive_rate_callback.c:54]: this is really clever - is there a reason you're not using float_s32_div? I see that you've gone so far as to comment it out below?

File: examples/ffva/src/usb/usb_audio.c [hunk 4]
```diff
@@ -53,8 +56,8 @@ uint8_t clkValid;
 audio_control_range_2_n_t(1) volumeRng[CFG_TUD_AUDIO_FUNC_1_N_CHANNELS_TX+1]; 			// Volume range state
 audio_control_range_4_n_t(1) sampleFreqRng; 						// Sample frequency range state
 
-static volatile bool mic_interface_open = false;
-static volatile bool spkr_interface_open = false;
+volatile bool mic_interface_open = false;
```
Comment (ACascarino) [inline examples/ffva/src/usb/usb_audio.c:59]: Ah, extern from here. Hmm.

**Review by ACascarino — APPROVED** [review]
Overall, just a couple of comments - could do with e.g. commenting from where variables are externed, but nothing here that would preclude a merge

File: examples/ffva/src/main.c (hunk 1 above)
Comment (shuchitak) [reply examples/ffva/src/main.c:315]: I meant to uncomment these before merging and forgot. Sorry, will do it now

File: examples/ffva/src/usb/adaptive_rate_adjust.c (hunk 2 above)
Comment (shuchitak) [reply examples/ffva/src/usb/adaptive_rate_adjust.c:42]: I'll add a comment. The set interface callback in usb_audio.c is the only place I can get the interface open information reliably from so doing an extern from there.

File: examples/ffva/src/usb/adaptive_rate_callback.c (hunk 3 above)
Comment (shuchitak) [reply examples/ffva/src/usb/adaptive_rate_callback.c:54]: The float_s32_div works on signed numbers. We can get a couple of extra bits of precision by doing unsigned division so decided to write my own.

**Pull request merged**